ir/src/symbol_table.rs

```diff
@@ -279,6 +312,30 @@ impl SymbolTable {
                 }
                 Ok(symbol_type)
             }
+            IdentifierType::TransitionVariable(transition_variable) => {
+                if let TransitionVariableType::Matrix(matrix) = transition_variable.value() {
+                    if matrix_access.row_idx() >= matrix.len() {
+                        return Err(SemanticError::matrix_access_out_of_bounds(
+                            matrix_access,
+                            matrix.len(),
+                            matrix[0].len(),
+                        ));
+                    }
+                    if matrix_access.col_idx() >= matrix[0].len() {
+                        return Err(SemanticError::matrix_access_out_of_bounds(
+                            matrix_access,
+                            matrix.len(),
+                            matrix[0].len(),
+                        ));
+                    }
+                    Ok(symbol_type)
```

Comment:
This whole section seems to be a duplication of the code in the `Constant` case above. Could we use a helper function instead?